Added two new features to facetoface

stgabhas opened the pull request:
Hello,
I have added two new features:
1) Disable Auto Enrollment of user xx days before a session starts, (when a user is on wait-list due to over booking)
2) Disable New Enrollment of user xx days before a session starts.

Thank you.

Commit:
Added two new features: (1)Disable Auto Enrollemnt xx days before a session starts. (2)Disable New Enrollment xx days before a session starts.

Commit:
removed unnecessary blocks/comments, white spaces, dubugging notices

stgabhas commented:
I have updated the code, as you commented. Please check and let me know.
Thank you.
Abhas

aly commented on lib.php:
```diff
@@ -561,13 +561,11 @@ function facetoface_update_calendar_entries($session, $facetoface = null){
  */
 function facetoface_update_attendees($session) {
     global $USER, $DB;
-
```
Re-add line deleted here

aly commented on lib.php:
```diff
@@ -561,13 +561,11 @@ function facetoface_update_calendar_entries($session, $facetoface = null){
 
     // Get course
     $course = $DB->get_record('course', array('id' => $facetoface->course));
-
```
Re-add new line

aly commented on lib.php:
```diff
@@ -598,7 +596,15 @@ function facetoface_update_attendees($session) {
                 }
             }
-
+            
```
This line has trailing whitespace

aly commented on lib.php:
```diff
@@ -598,7 +596,15 @@ function facetoface_update_attendees($session) {
+            $start = $DB->get_record('facetoface_sessions_dates', array('sessionid'=>$session->id)); 
+            $starttime = new DateTime(date('c', $start->timestart)); 
+            $interval = $starttime->diff($currenttime); 
+			$diffdays =  $interval->format('%a'); 
```
The indentation of these lines is wrong, I suspect its because a tab has been used rather than spaces (Moodle coding style specifies that the indenting must be 4 spaces for each level of indentation)

aly commented on lib.php:
```diff
@@ -606,9 +612,9 @@ function facetoface_update_attendees($session) {
+                    if (($user->statuscode == MDL_F2F_STATUS_WAITLISTED)&($diffdays>$disableddays)) {
 
-                        if (!facetoface_user_signup($session, $facetoface, $course, $user->discountcode, $user->notificationtype, MDL_F2F_STATUS_BOOKED, $user->id)) {
+						if (!facetoface_user_signup($session, $facetoface, $course, $user->discountcode, $user->notificationtype, MDL_F2F_STATUS_BOOKED, $user->id)) {
```
Indentation is broken

aly commented on lib.php:
```diff
@@ -1636,7 +1643,6 @@ function facetoface_write_activity_attendance(&$worksheet, $startingrow, $faceto
             }
         }
     }
-
```
re-add removed newline

aly commented on renderer.php:
```diff
@@ -138,8 +140,27 @@ public function print_session_list_table($customfields, $sessions, $viewattendee
-                $options .= html_writer::link('signup.php?s='.$session->id.'&backtoallsessions='.$session->facetoface, get_string('signup', 'facetoface'));
-            }
+		$currenttime = new DateTime(date('c', time()));
+                $start = $DB->get_record('facetoface_sessions_dates', array('sessionid'=>$session->id));
```
Database calls should not be make in the renderer. All data that is needed in the renderer should be fetched then passed into the rendering function

stgabhas replied:
To fix this, can I write the code below in view.php

```
143                 $currenttime = new DateTime(date('c', time()));
144                 $start = $DB->get_record('facetoface_sessions_dates', array('sessionid'=>$session->id));
145                 $starttime = new DateTime(date('c', $start->timestart));
146                 $interval = $starttime->diff($currenttime);
147                 $diffdays =  $interval->format('%a');
148                 $disableddaysDB = $DB->get_record('facetoface_sessions', array('id'=>$session->id));
149                 $disableddays = $disableddaysDB->disablenewenrolldays;
150                 $disableoption = $disableddaysDB->disableoption;
```

and pass $disableoption as an argument in (print_session_list_table) that is being called in line 182 and 192 in view.php

aly replied:
I've had a closer look at the renderer and the date information is already available. If you look at the code around like 68 or renderer.php you will see that the code loops through all the dates for the session. You should be able to use this information to achieve what you need to. Note that the session may contain multiple dates so I'm assuming you will want to compare again the earliest one.

aly commented on renderer.php:
```diff
@@ -138,8 +140,27 @@ public function print_session_list_table($customfields, $sessions, $viewattendee
+                {
+                        if($diffdays > $disableddays)
+                        {
+                                $options .= html_writer::link('signup.php?s='.$session->id.'&backtoallsessions='.$session->facetoface, get_string('signup', 'facetoface'));
```
When creating a link, use the moodle_url class to create the url

new moodle_url('signup.php', array('s' => $session->id, 'backtoallsessions' => $session->facetoface)

aly commented on renderer.php:
```diff
@@ -138,8 +140,27 @@ public function print_session_list_table($customfields, $sessions, $viewattendee
+                }
+                else
+                {
+                        $options .= html_writer::link('signup.php?s='.$session->id.'&backtoallsessions='.$session->facetoface, get_string('signup', 'facetoface'));
```
use moodle_url here too

aly commented on renderer.php:
```diff
@@ -163,7 +184,7 @@ public function print_session_list_table($customfields, $sessions, $viewattendee
 
         $output .= html_writer::table($table);
-
+       
```
Remove added whitespace on this line

aly commented on session_form.php:
```diff
@@ -125,6 +125,23 @@ function definition() {
+        $mform->setDefault('disableoption', 0);
+        $mform->addHelpButton('disableoption', 'disableoption', 'facetoface');
+        
+        
```
Remove extra new line

aly commented on session_form.php:
```diff
@@ -125,6 +125,23 @@ function definition() {
+        $mform->disabledIf('disablenewenrolldays', 'disableoption');
+        $mform->addHelpButton('disablenewenrolldays', 'disablenewenrolldays', 'facetoface');
+
+
```
Remove extra new line

aly commented on sessions.php:
```diff
@@ -122,6 +122,15 @@
         $fromform->discountcost = 0;
     }
 
+    if(empty($fromform->disablenewenrolldays))
```
Fix layout of this if statement according to previous comment

aly commented on sessions.php:
```diff
@@ -228,6 +241,11 @@
     $toform->discountcost = $session->discountcost;
 
+    $toform->details = $session->details;
+    //Am
```
Remove this comment

aly commented on view.php:
```diff
@@ -185,11 +185,11 @@ function print_session_list($courseid, $facetofaceid, $location) {
         echo html_writer::tag('p', html_writer::link(new moodle_url('sessions.php', array('f' => $facetofaceid)), get_string('addsession', 'facetoface')));
     }
-
+			
```
Remove added whitespace

aly commented on view.php:
```diff
@@ -185,11 +185,11 @@ function print_session_list($courseid, $facetofaceid, $location) {
     if (!empty($previousarray)) {
         echo $OUTPUT->heading(get_string('previoussessions', 'facetoface'));
-        echo $f2f_renderer->print_session_list_table($customfields, $previousarray, $viewattendees, $editsessions);
+	echo $f2f_renderer->print_session_list_table($customfields, $previousarray, $viewattendees, $editsessions);
```
Fix indenting, in this case should be 4 spaces (no tabs)

aly commented on db/upgrade.php:
```diff
@@ -697,5 +697,47 @@ function xmldb_facetoface_upgrade($oldversion=0) {
         upgrade_mod_savepoint(true, 2013010400, 'facetoface');
     }
 
+    if ($oldversion < 2012051100) {
```
Change this one to 2013010401 the format for this number is YYYYMMDDXX there is documentation on version number in Moodle here http://docs.moodle.org/dev/version.php

aly commented on db/upgrade.php:
```diff
@@ -697,5 +697,47 @@ function xmldb_facetoface_upgrade($oldversion=0) {
+        upgrade_mod_savepoint(true, 2012051100, 'facetoface');
+    }
+
+    if ($oldversion < 2012051200) {
```
Make these 3 changes upgrades into one

aly commented on mod_form.php:
```diff
@@ -136,6 +136,19 @@ function definition()
+        $mform->addHelpButton('disableautoenroll', 'disableautoenroll', 'facetoface');
+        
+        $disableperiod = array();
+        for ($i=0; $i<=365; $i += 1) {
```
I'm hesitant about using zero as it might be confusing to users. Also going all the way up to 365 day before the session is quite a long way. Maybe it would be good to make the number lower (60 or 90 days). I would ideally like to have a selector which allowed selection of "3 months" or "4 weeks".

aly commented:
Few more comments, once those are fixed I'll test it out :-)

Commit:
Fixed the comments/issues, white spaces, indentation; except for the Database calls that is made in the renderer

stgabhas commented:
I have pushed a new update as according to your input. (Except for the calling of DB in renderer).
About the Day 0 in the enrollment, I have added an explanation that Day 0 means the same day the session starts.
Waiting for your comments.
Thank you.

aly commented on db/upgrade.php:
```diff
@@ -697,5 +697,39 @@ function xmldb_facetoface_upgrade($oldversion=0) {
+        $table = new xmldb_table('facetoface');
+        $field = new xmldb_field('disablewithindays', XMLDB_TYPE_INTEGER, '3', null, XMLDB_NOTNULL, null, '0', 'approvalreqd');
+
+
```
Remove extra new line

aly commented on db/upgrade.php:
```diff
@@ -697,5 +697,39 @@ function xmldb_facetoface_upgrade($oldversion=0) {
+        }
+
+        // facetoface savepoint reached
+        upgrade_mod_savepoint(true, 2012051200, 'facetoface');
```
Remove this line and the comment and newline above so there is only only new line between this and the next line of code. This will break the upgrade

aly commented on db/upgrade.php:
```diff
@@ -697,5 +697,39 @@ function xmldb_facetoface_upgrade($oldversion=0) {
+        }
+
+        // facetoface savepoint reached
+        upgrade_mod_savepoint(true, 2012053002, 'facetoface');
```
This version number needs to match the version number at the start of the upgrade block

aly commented on lib.php:
```diff
@@ -598,6 +598,14 @@ function facetoface_update_attendees($session) {
             }
+ 
+            $currenttime = new DateTime(date('c', time())); 
+            $start = $DB->get_record('facetoface_sessions_dates', array('sessionid'=>$session->id)); 
```
There is a function in lib.php called "facetoface_get_session_dates" with a parameter of $sessionid. Use this instead of making a database call

aly commented on lib.php:
```diff
@@ -598,6 +598,14 @@ function facetoface_update_attendees($session) {
+ 
+            $currenttime = new DateTime(date('c', time())); 
+            $start = $DB->get_record('facetoface_sessions_dates', array('sessionid'=>$session->id)); 
+            $starttime = new DateTime(date('c', $start->timestart)); 
```
A session may have many dates, this code will not work correctly if a session has more than one date

aly commented on lib.php:
```diff
@@ -606,8 +614,7 @@ function facetoface_update_attendees($session) {
 
-                    if ($user->statuscode == MDL_F2F_STATUS_WAITLISTED) {
-
+                    if (($user->statuscode == MDL_F2F_STATUS_WAITLISTED)&($diffdays>$disableddays)) {
```
in this you should use "&&" instead of "&". The way it is currently done it is performing a bitwise operation rather than a logical comparison. Also put a space on either side of the "&&" after changing

aly commented on lib.php:
```diff
@@ -1447,7 +1453,8 @@ function facetoface_write_activity_attendance(&$worksheet, $startingrow, $faceto
               AND d.sessionid = s.id
                    $locationcondition
-                   ORDER BY s.datetimeknown, d.timestart";
+                   ORDER BY s.datetimeknown, d.timestart"; 
```
This line has an extra space at the end which needs to be removed

aly commented on renderer.php:
```diff
@@ -8,7 +8,8 @@ class mod_facetoface_renderer extends plugin_renderer_base {
     public function print_session_list_table($customfields, $sessions, $viewattendees, $editsessions) {
-        $output = '';
+        global $DB;
+	$output = '';
```
fix indenting

aly commented on renderer.php:
```diff
@@ -32,6 +33,7 @@ public function print_session_list_table($customfields, $sessions, $viewattendee
         $table = new html_table();
         $table->summary = get_string('previoussessionslist', 'facetoface');
         $table->head = $tableheader;
+	$table->width = '100%';
```
Please add styling by adding a class to the table and adding appropriate css to styles.css

stgabhas replied:
To perform this:
I removed the line: $table->width = '100%';
I added the line: $table->attributes['class'] = 'sessionlisttable'; on renderer.php
And I added the line: .sessionlisttable { width: 100%; } on styles.css

But the width of the table is Not 100% on the output page. Can you please check what I did wrong. I looked into moodle api and into moodle forums and I see that is how it is done, but still I did not get the desired output.

aly commented on renderer.php:
```diff
@@ -138,8 +140,25 @@ public function print_session_list_table($customfields, $sessions, $viewattendee
+                    if($diffdays > $disableddays) {
+                        $options .= new moodle_url('signup.php', array('s' => $session->id, 'backtoallsessions' => $session->facetoface), get_string('signup', 'facetoface'));
+                    }
+                }
```
need to move else on the same line as this bracket as shown in Moodle coding style guidelines

aly commented on renderer.php:
```diff
@@ -138,8 +140,25 @@ public function print_session_list_table($customfields, $sessions, $viewattendee
+                $disableddays = $disableddaysDB->disablenewenrolldays;
+                $disableoption = $disableddaysDB->disableoption;
+                
+                if($disableoption) {
```
add space after if and before (

aly commented on sessions.php:
```diff
@@ -122,6 +122,15 @@
+        $fromform->disablenewenrolldays = 0;
+    }
+
+    if (empty($fromform->disableoption))
```
move bracket onto same line as if statement

aly commented on lib.php:
```diff
@@ -1447,7 +1453,8 @@ function facetoface_write_activity_attendance(&$worksheet, $startingrow, $faceto
                    $locationcondition
-                   ORDER BY s.datetimeknown, d.timestart";
+                   ORDER BY s.datetimeknown, d.timestart"; 
+
```
remove extra new line

aly commented on db/install.xml:
```diff
@@ -34,7 +34,8 @@
         <FIELD NAME="approvalreqd" TYPE="int" LENGTH="1" NOTNULL="true" UNSIGNED="true" DEFAULT="0" SEQUENCE="false" PREVIOUS="showoncalendar" NEXT="usercalentry"/>
-        <FIELD NAME="usercalentry" TYPE="int" LENGTH="1" NOTNULL="true" UNSIGNED="true" DEFAULT="1" SEQUENCE="false" PREVIOUS="approvalreqd"/>
+        <FIELD NAME="usercalentry" TYPE="int" LENGTH="1" NOTNULL="true" UNSIGNED="true" DEFAULT="1" SEQUENCE="false" PREVIOUS="approvalreqd" NEXT="disablewithindays"/>
+	<FIELD NAME="disablewithindays" TYPE="int" LENGTH="3" NOTNULL="true" UNSIGNED="true" DEFAULT="0" SEQUENCE="false" PREVIOUS="usercalentry"/>
```
fix indenting of this line

aly commented on db/install.xml:
```diff
@@ -67,8 +68,10 @@
-      </FIELDS>
+        <FIELD NAME="timemodified" TYPE="int" LENGTH="20" NOTNULL="true" UNSIGNED="true" DEFAULT="0" SEQUENCE="false" PREVIOUS="timecreated" NEXT="disablenewenrolldays"/>
+        <FIELD NAME="disablenewenrolldays" TYPE="int" LENGTH="3" NOTNULL="true" UNSIGNED="true" DEFAULT="0" SEQUENCE="false" PREVIOUS="timemodified" NEXT="disableoption"/>
+        <FIELD NAME="disableoption" TYPE="int" LENGTH="1" NOTNULL="false" UNSIGNED="true" SEQUENCE="false" PREVIOUS="disablenewenrolldays"/>
```
The name of this field does not describe what it does. A better name might be "disablesignup".

aly commented on renderer.php:
```diff
@@ -138,8 +140,25 @@ public function print_session_list_table($customfields, $sessions, $viewattendee
+                $starttime = new DateTime(date('c', $start->timestart));
+                $interval = $starttime->diff($currenttime);
+                $diffdays =  $interval->format('%a');
+                $disableddaysDB = $DB->get_record('facetoface_sessions', array('id'=>$session->id));
```
This DB call is a little harder to remove but you should be able to pass through the information you need from where print_session_list_table is called. In view.php

aly commented:
Hopefully my comments about the Database calls is helpful. If you have any questions let me know.

Thanks,
Alastair

Commit:
Fixed all the issues according to your last comment regarding white space, extra lines, and indentation. The css 100% width on styles.php is not working, and I am still working on the database calls mentioned in the last comment

Commit:
Removed some database call from renderer.php

Commit:
Removed a database call on lib.php. Got the date from facetoface_get_session_dates function

stgabhas commented:
I have updated the code as instructed. And I have removed a couple of database calls from renderer.php and lib.php. Please refer to todays commit.
Thank you.

stgabhas commented:
The issue with 100% width on styles.css has been fixed as well.

Commit:
custom facetoface for moodle 2.5.1 for servicetraining

Commit:
mod_form.php: add set type for thirdparty parameter

Commit:
session_form.php: add set type for session form options

Commit:
lib.php: use sql_compare_text when deleting session calendar events

Commit:
cancelsignup_form.php: add set type for cancelsignup_form.php

Commit:
mod_form.php: use correct type PARAM_EMAIL for thirdparty parameter

Commit:
lang/en: remove broken <p> tag from allowoverbooking_desc

Commit:
misc/forms: ensure setType is set on form elements

Commit:
version.php: bumping to version 2.2.4.1

Commit:
version.php: update phpdoc

Commit:
settings.php: fix get system roles

Commit:
2.6: add get_all_user_name_fields calls where user firstname/lastname is
called

Commit:
version.php bump for version 2.3.0 (Moodle 2.6)

Commit:
ChangeLog.txt: update changelog

Commit:
lib.php: count ID value not all records

Commit:
lib.php: fix bug in cancelled users GROUP BY and candidate status check

Commit:
Bumping to version 2.3.1

Commit:
README: update readme with correct maintainer information

Commit:
lib.php: fix bug with missing u.id in GROUP BY clause

Commit:
version.php: bump for development versioning

stgabhas committed:
changes from catalyst Stacey Walker
